pkg/artifactory/resource/security/resource_artifactory_scoped_token.go

```diff
@@ -0,0 +1,285 @@
+package security
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-resty/resty/v2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/jfrog/terraform-provider-shared/util"
+	"github.com/jfrog/terraform-provider-shared/validator"
+)
+
+type AccessTokenPostRequest struct {
+	GrantType   string `json:"grant_type"`
+	Username    string `json:"username,omitempty"`
+	Scope       string `json:"scope,omitempty"`
+	ExpiresIn   int    `json:"expires_in,omitempty"`
+	Refreshable bool   `json:"refreshable"`
+	Description string `json:"description"`
+	Audience    string `json:"audience,omitempty"`
+}
+
+type AccessTokenPostResponse struct {
+	TokenId     string `json:"token_id"`
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
+	Scope       string `json:"scope"`
+	TokenType   string `json:"token_type"`
+}
+
+type AccessTokenGet struct {
+	TokenId     string `json:"token_id"`
+	Subject     string `json:"subject"`
+	Expiry      int    `json:"expiry"`
+	IssuedAt    int    `json:"issued_at"`
+	Issuer      string `json:"issuer"`
+	Description string `json:"description"`
+	Refreshable bool   `json:"refreshable"`
+}
+
+func (a AccessTokenPostResponse) Id() string {
+	return a.TokenId
+}
+
+const accessTokensUrl = "access/api/v1/tokens"
+const accessTokenUrl = accessTokensUrl + "/{id}"
+
+func ResourceArtifactoryScopedToken() *schema.Resource {
+
+	var unpackAccessTokenPostRequest = func(data *schema.ResourceData) (*AccessTokenPostRequest, error) {
+		d := &util.ResourceData{data}
+
+		scopes := d.GetSet("scopes")
+		scopesString := strings.Join(scopes, " ") // Join slice into space-separated string
+		if len(scopesString) > 500 {
+			return nil, fmt.Errorf("Total combined length of scopes field exceeds 500 characters: %s", scopesString)
+		}
+
+		audiences := d.GetSet("audiences")
+		audiencesString := strings.Join(audiences, " ") // Join slice into space-separated string
+		if len(audiencesString) > 255 {
+			return nil, fmt.Errorf("Total combined length of audences field exceeds 255 characters: %s", audiencesString)
+		}
+
+		accessToken := AccessTokenPostRequest{
+			Username:    d.GetString("username", false),
+			Scope:       scopesString,
+			ExpiresIn:   d.GetInt("expires_in", false),
+			Refreshable: d.GetBool("refreshable", false),
+			Description: d.GetString("description", false),
+			Audience:    audiencesString,
+		}
+
+		return &accessToken, nil
+	}
+
+	var packAccessToken = func(d *schema.ResourceData, accessToken AccessTokenGet) diag.Diagnostics {
+		setValue := util.MkLens(d)
+
+		var errors []error
+
+		errors = append(errors, setValue("subject", accessToken.Subject)...)
+		errors = append(errors, setValue("expiry", accessToken.Expiry)...)
+		errors = append(errors, setValue("issued_at", accessToken.IssuedAt)...)
```

Comment:
If I am not mistaken, you shouldn't need to do this `append` - Internally it uses a closure and it always contains all errors